MinGW `-mwindows` flag support

I have this simple win32 program:

```cpp
#include <windows.h>

int WINAPI WinMain(HINSTANCE hInst, HINSTANCE hInstPrev, PSTR cmdline, int cmdshow)
{
    return MessageBox(NULL, "Hello, worl!", "Caption", 0);
}
```

Cross-compiling it from my Mac just works.

```
$ zig cc -o main.exe main.cpp -target x86_64-windows
```

(Which never ceases to amaze me.)

When starting the .exe from the Windows File Explorer, a terminal window shows up:

<img width="323" alt="image" src="https://user-images.githubusercontent.com/76071/161837815-cb9211e0-e849-4f5d-b056-7d2d4f30f91f.png">

There are workarounds to make the terminal flash just briefly but that is still not ideal.

I found out there's a `-mwindows` flag I can pass when cross-compiling with MinGW:

```
$ x86_64-w64-mingw32-g++ -o main.exe main.cpp -mwindows
```

Without it, I have the same behaviour like with Zig, but _with_ it, there's no terminal window.

I wasn't able to determine what exactly that flag does. From little I was able to find it seems by default programs are built in the "console" mode and in this particular case we want the "GUI" mode.

Per:

```
$ x86_64-w64-mingw32-g++ --help | grep -- -m
Options starting with -g, -f, -m, -O, -W, or --param are automatically passed on to the various sub-processes invoked by x86_64-w64-mingw32-g++.
```

Perhaps it's:

```
$ x86_64-w64-mingw32-ld --help | grep -- -m
...
-m EMULATION                Set emulation
...
```

EDIT: it's not an ld flag. Possible options for `-m` are: `-mwindows`, `-mconsole`, and `-mdll`.

In any case, if it was desirable to expose this flag somehow, I'd be happy to look into it. If so, any pointers would be appreciated.

@Ristovski excellent, thank you.

Unfortunately I'm getting this error:

```
$ zig build-exe main.cpp -target x86_64-windows
error(compilation): clang failed with stderr: main.cpp:1:10: fatal error: 'windows.h' file not found

main.cpp:1:1: error: unable to build C object: clang exited with code 1
```

It looks like Zig is maybe not running in cross-compilation mode?

--- a/src/main.zig
+++ b/src/main.zig
@@ -849,36 +849,7 @@ fn buildOutputType(
                         const next_arg = args_iter.next() orelse {
                             fatal("expected parameter after {s}", .{arg});
                         };
-                        if (mem.eql(u8, next_arg, "console")) {
-                            subsystem = .Console;
-                        } else if (mem.eql(u8, next_arg, "windows")) {
-                            subsystem = .Windows;
-                        } else if (mem.eql(u8, next_arg, "posix")) {
-                            subsystem = .Posix;
-                        } else if (mem.eql(u8, next_arg, "native")) {
-                            subsystem = .Native;
-                        } else if (mem.eql(u8, next_arg, "efi_application")) {
-                            subsystem = .EfiApplication;
-                        } else if (mem.eql(u8, next_arg, "efi_boot_service_driver")) {
-                            subsystem = .EfiBootServiceDriver;
-                        } else if (mem.eql(u8, next_arg, "efi_rom")) {
-                            subsystem = .EfiRom;
-                        } else if (mem.eql(u8, next_arg, "efi_runtime_driver")) {
-                            subsystem = .EfiRuntimeDriver;
-                        } else {
-                            fatal("invalid: --subsystem: '{s}'. Options are:\n{s}", .{
-                                next_arg,
-                                \\  console
-                                \\  windows
-                                \\  posix
-                                \\  native
-                                \\  efi_application
-                                \\  efi_boot_service_driver
-                                \\  efi_rom
-                                \\  efi_runtime_driver
-                                \\
-                            });
-                        }
+                        subsystem = try parseSubSystem(next_arg);
                     } else if (mem.eql(u8, arg, "-O")) {
                         optimize_mode_string = args_iter.next() orelse {
                             fatal("expected parameter after {s}", .{arg});
@@ -1467,6 +1438,11 @@ fn buildOutputType(
                                 mem.eql(u8, linker_arg, "-static"))
                             {
                                 force_static_libs = true;
+                            } else if (mem.eql(u8, linker_arg, "--subsystem")) {
+                                const next_arg = split_it.next() orelse {
+                                    fatal("expected parameter after {s}", .{linker_arg});
+                                };
+                                subsystem = try parseSubSystem(next_arg);
                             } else {
                                 try linker_args.append(linker_arg);
                             }
@@ -5130,5 +5106,38 @@ fn warnAboutForeignBinaries(
                 host_name, foreign_name, tip_suffix,
             });
         },
+    }
+}
+
+fn parseSubSystem(next_arg: []const u8) !std.Target.SubSystem {
+    if (mem.eql(u8, next_arg, "console")) {
+        return .Console;
+    } else if (mem.eql(u8, next_arg, "windows")) {
+        return .Windows;
+    } else if (mem.eql(u8, next_arg, "posix")) {
+        return .Posix;
+    } else if (mem.eql(u8, next_arg, "native")) {
+        return .Native;
+    } else if (mem.eql(u8, next_arg, "efi_application")) {
+        return .EfiApplication;
+    } else if (mem.eql(u8, next_arg, "efi_boot_service_driver")) {
+        return .EfiBootServiceDriver;
+    } else if (mem.eql(u8, next_arg, "efi_rom")) {
+        return .EfiRom;
+    } else if (mem.eql(u8, next_arg, "efi_runtime_driver")) {
+        return .EfiRuntimeDriver;
+    } else {
+        fatal("invalid: --subsystem: '{s}'. Options are:\n{s}", .{
+            next_arg,
+            \\  console
+            \\  windows
+            \\  posix
+            \\  native
+            \\  efi_application
+            \\  efi_boot_service_driver
+            \\  efi_rom
+            \\  efi_runtime_driver
+            \\
+        });
     }
 }

--- a/src/main.zig
+++ b/src/main.zig
@@ -1438,11 +1438,6 @@ fn buildOutputType(
                                 mem.eql(u8, linker_arg, "-static"))
                             {
                                 force_static_libs = true;
-                            } else if (mem.eql(u8, linker_arg, "--subsystem")) {
-                                const next_arg = split_it.next() orelse {
-                                    fatal("expected parameter after {s}", .{linker_arg});
-                                };
-                                subsystem = try parseSubSystem(next_arg);
                             } else {
                                 try linker_args.append(linker_arg);
                             }
@@ -1586,6 +1581,12 @@ fn buildOutputType(
                         fatal("expected linker arg after '{s}'", .{arg});
                     }
                     try rpath_list.append(linker_args.items[i]);
+                } else if (mem.eql(u8, arg, "--subsystem")) {
+                    i += 1;
+                    if (i >= linker_args.items.len) {
+                        fatal("expected linker arg after '{s}'", .{arg});
+                    }
+                    subsystem = try parseSubSystem(linker_args.items[i]);
                 } else if (mem.eql(u8, arg, "-I") or
                     mem.eql(u8, arg, "--dynamic-linker") or
                     mem.eql(u8, arg, "-dynamic-linker"))

